modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Service.java

```diff
@@ -361,10 +363,19 @@ static class CustomWebIdentityTokenCredentialsProvider implements AWSCredentials
             );
             AWSSecurityTokenServiceClientBuilder stsClientBuilder = AWSSecurityTokenServiceClient.builder();
 
-            // Custom system property used for specifying a mocked version of the STS for testing
-            String customStsEndpoint = jvmEnvironment.getProperty("com.amazonaws.sdk.stsMetadataServiceEndpointOverride", STS_HOSTNAME);
-            // Set the region explicitly via the endpoint URL, so the AWS SDK doesn't make any guesses internally.
-            stsClientBuilder.withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(customStsEndpoint, null));
+            // AWS_REGION should be injected by the EKS pod identity webhook:
+            // https://github.com/aws/amazon-eks-pod-identity-webhook/pull/41
+            stsRegion = systemEnvironment.getEnv(SDKGlobalConfiguration.AWS_REGION_ENV_VAR);
+            // Check if we need to use regional STS endpoints
+            // https://docs.aws.amazon.com/sdkref/latest/guide/feature-sts-regionalized-endpoints.html
+            if ("regional".equalsIgnoreCase(systemEnvironment.getEnv("AWS_STS_REGIONAL_ENDPOINTS")) && stsRegion != null) {
+                stsClientBuilder.withRegion(stsRegion);
+            } else {
+                // Custom system property used for specifying a mocked version of the STS for testing
+                String customStsEndpoint = jvmEnvironment.getProperty("com.amazonaws.sdk.stsMetadataServiceEndpointOverride", STS_HOSTNAME);
+                // Set the region explicitly via the endpoint URL, so the AWS SDK doesn't make any guesses internally.
+                stsClientBuilder.withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(customStsEndpoint, null));
+            }
```

Comment:
Should we validate and report error if `AWS_STS_REGIONAL_ENDPOINTS` is set to `regional` but `AWS_REGION_ENV_VAR` has no value or if `AWS_REGION_ENV_VAR` has value but `AWS_STS_REGIONAL_ENDPOINTS` is not `regional` instead of just falling to the `else` branch?

Reply:
Good point! 

I've added logging for the case `AWS_STS_REGIONAL_ENDPOINTS` is to `regional`, but `AWS_REGION` is not set for some reason. The opposite case is not needed, I think. `AWS_REGION` can be automatically injected, but the user can prefer to use the global endpoint or a set custom one via the `com.amazonaws.sdk.stsMetadataServiceEndpointOverride` property. Setting `AWS_STS_REGIONAL_ENDPOINTS` to `regional` should indicate the intent to use regional endpoints.